### PR TITLE
Parse statements in Flow

### LINKED ISSUE #735: Parse statements in Flow
Contributes to #584 
- [x] type alias statement, with `opaque`, with supertype
- [x] import statements
- [x] export statements
- [ ] declare type statement

### PR DESCRIPTION
**Pull request by vilchik-elena**
Fixes #735

### CODE REVIEW

File: javascript-frontend/src/test/java/org/sonar/javascript/tree/impl/flow/FlowTypeAliasStatementTreeModelTest.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    assertThat(tree.opaqueToken()).isNull();
+    assertThat(tree.typeToken().text()).isEqualTo("type");
+    assertThat(tree.typeAlias().name()).isEqualTo("myAlias");
+    assertThat(tree.type().is(Kind.FLOW_OPTIONAL_TYPE)).isTrue();
+    assertThat(tree.semicolonToken().text()).isEqualTo(";");
+  }
+
+  @Ignore
+  @Test
+  public void without_semicolon() throws Exception {
```
Comment (SonarTech) [inline javascript-frontend/src/test/java/org/sonar/javascript/tree/impl/flow/FlowTypeAliasStatementTreeModelTest.java]: MAJOR Fix or remove this skipped unit test [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1607)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: javascript-frontend/src/main/java/org/sonar/javascript/parser/JavaScriptGrammar.java [hunk 2]
```diff
@@ -1769,6 +1773,33 @@ public FlowOptionalBindingElementTree FLOW_OPTIONAL_BINDING_ELEMENT() {
...
+        FLOW_TYPE(),
+        b.token(EcmaScriptLexer.EOS)));
+  }
+
+  public InternalSyntaxToken FLOW_TYPE_KEYWORD() {
+    return b.<InternalSyntaxToken>nonterminal()
+      .is(b.firstOf(b.token(JavaScriptKeyword.TYPEOF), b.token(EcmaScriptLexer.TYPE)));
+  }
+
+  // TODO: temp impl for export declaration
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/parser/JavaScriptGrammar.java]: INFO Complete the task associated to this TODO comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)

File: javascript-frontend/src/main/java/org/sonar/javascript/tree/impl/flow/FlowInterfaceDeclarationTreeImpl.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  }
+
+  @Override
+  public Iterator<Tree> childrenIterator() {
+    return Iterators.forArray(interfaceToken, name, leftCurlyBraceToken , rightCurlyBraceToken);
+  }
+
+  @Override
+  public void accept(DoubleDispatchVisitor visitor) {
+    // TODO
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/tree/impl/flow/FlowInterfaceDeclarationTreeImpl.java]: INFO Complete the task associated to this TODO comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

File: javascript-frontend/src/main/java/org/sonar/javascript/parser/TreeFactory.java [hunk 4]
```diff
@@ -1164,10 +1183,10 @@ public NameSpaceExportDeclarationTree namespaceExportDeclaration(
       asToken = nameSpaceExport.get().first;
       synonymIdentifier = nameSpaceExport.get().second;
     }
-    return new NameSpaceExportDeclarationTreeImpl(exportToken, starToken, asToken, synonymIdentifier, fromClause, nullableSemicolonToken(semicolonToken));
+    return new NameSpaceExportDeclarationTreeImpl(exportToken, flowTypeKeywordToken.orNull(), starToken, asToken, synonymIdentifier, fromClause, nullableSemicolonToken(semicolonToken));
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/parser/TreeFactory.java]: MINOR Split this 185 characters long line (which is greater than 180 authorized). [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00103)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![MINOR][MINOR] 1 minor
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: javascript-frontend/src/main/java/org/sonar/javascript/parser/JavaScriptGrammar.java [hunk 5]
```diff
@@ -1271,53 +1272,40 @@ public ExportDeclarationTree EXPORT_DECLARATION() {
...
-        )
-      ));
+      .is(b.firstOf(
+        f.importClauseWithTwoParts(BINDING_IDENTIFIER(), b.token(JavaScriptPunctuator.COMMA), b.firstOf(NAMESPACE_IMPORT(), NAMED_IMPORTS())),
+        f.importClause(BINDING_IDENTIFIER()),
+        f.importClause(NAMESPACE_IMPORT()),
+        f.importClause(NAMED_IMPORTS())));
   }
 
   public DeclarationTree IMPORT_DECLARATION() {
```
Comment (inverno) [inline javascript-frontend/src/main/java/org/sonar/javascript/parser/JavaScriptGrammar.java]: Consider pushing this and the import clause up, before `NAMED_IMPORTS`

File: javascript-frontend/src/main/java/org/sonar/javascript/parser/JavaScriptGrammar.java [hunk 6]
```diff
@@ -1286,7 +1286,9 @@ public NamedImportExportClauseTree NAMED_IMPORTS() {
   public SpecifierTree IMPORT_SPECIFIER() {
     return b.<SpecifierTree>nonterminal(Kind.IMPORT_SPECIFIER)
       .is(b.firstOf(
+        f.importSpecifier(FLOW_TYPE_KEYWORD(), IDENTIFIER_NAME(), b.token(EcmaScriptLexer.AS), BINDING_IDENTIFIER()),
```
Comment (inverno) [inline javascript-frontend/src/main/java/org/sonar/javascript/parser/JavaScriptGrammar.java]: try with optional

Comment (inverno) [reply javascript-frontend/src/main/java/org/sonar/javascript/parser/JavaScriptGrammar.java]: Nope. Does not work

File: javascript-frontend/src/main/java/org/sonar/javascript/parser/TreeFactory.java [hunk 7]
```diff
@@ -1790,6 +1825,11 @@ public FlowFunctionTypeParameterClauseTree flowFunctionTypeParameterClause(
     return new FlowFunctionTypeParameterClauseTreeImpl(lParenthesis, new SeparatedListImpl(parameters, ImmutableList.of()), rParenthesis);
   }
 
+  // TODO use this when implementing single-parameter function type without parenthesis
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/parser/TreeFactory.java]: INFO Complete the task associated to this TODO comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)

File: javascript-frontend/src/main/java/org/sonar/javascript/tree/impl/flow/FlowInterfaceDeclarationTreeImpl.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+  }
+
+  @Override
+  public Iterator<Tree> childrenIterator() {
+    return Iterators.forArray(interfaceToken, name, leftCurlyBraceToken , rightCurlyBraceToken);
+  }
+
+  @Override
+  public void accept(DoubleDispatchVisitor visitor) {
+    // TODO when interface is filled
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/tree/impl/flow/FlowInterfaceDeclarationTreeImpl.java]: INFO Complete the task associated to this TODO comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* ![MINOR][MINOR] 1 minor
* ![INFO][INFO] 3 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**